### PR TITLE
Ignore staking queue

### PR DESCRIPTION
**Pull request by mariusmihaic**
- Added staking v4 flag in `staking.go` which has been used to deactivate functions which will be obsolete since staking v4 is enabled. It has been mostly used to _"inhibit"_ staking queue/ waiting list.
- Added unit tests

### CODE REVIEW

File: vm/systemSmartContracts/staking.go [hunk 1]
```diff
@@ -1289,6 +1311,12 @@ func (s *stakingSC) isNodeJailedOrWithBadRating(registrationData *StakedDataV2_0
 }
 
 func (s *stakingSC) getWaitingListIndex(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if s.flagStakingV4.IsSet() {
+		s.eei.AddReturnMessage(vm.ErrWaitingListDisabled.Error())
+		s.eei.Finish([]byte{0})
+
+		return vmcommon.Ok
```
Comment (sasurobert) [inline vm/systemSmartContracts/staking.go:1318]: if you have return message than return vmcommon.UserError, and do not add to Finish anything. as it is ignored

File: vm/systemSmartContracts/staking.go [hunk 2]
```diff
@@ -1353,6 +1381,13 @@ func (s *stakingSC) getWaitingListIndex(args *vmcommon.ContractCallInput) vmcomm
 }
 
 func (s *stakingSC) getWaitingListSize(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if s.flagStakingV4.IsSet() {
+		s.eei.AddReturnMessage(vm.ErrWaitingListDisabled.Error())
+		s.eei.Finish([]byte{0})
```
Comment (sasurobert) [inline vm/systemSmartContracts/staking.go:1386]: if you have return message than return vmcommon.UserError, and do not add to Finish anything. as it is ignored

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/staking.go (hunk 1 above)
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go]: Thanks for heads up, refactored it

File: vm/systemSmartContracts/staking.go (hunk 2 above)
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go]: Thanks for heads up, refactored it

File: vm/systemSmartContracts/staking_test.go [hunk 3]
```diff
@@ -3283,6 +3339,87 @@ func TestStakingSc_fixMissingNodeAddOneNodeOnly(t *testing.T) {
...
+
+	args := createMockStakingScArguments()
+	eei, _ := NewVMContext(&mock.BlockChainHookStub{}, hooks.NewVMCryptoHook(), &mock.ArgumentParserMock{}, &stateMock.AccountsStub{}, &mock.RaterMock{})
+	args.Eei = eei
+
+	stakingSmartContract, _ := NewStakingSmartContract(args)
+	stakingSmartContract.EpochConfirmed(args.EpochConfig.EnableEpochs.StakingV4EnableEpoch, 0)
+
+	arguments := CreateVmContractCallInput()
+	arguments.Arguments = [][]byte{}
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking_test.go:3353]: Right, can be removed. I've put it there so I don't pass nil pointers

File: vm/systemSmartContracts/staking_test.go [hunk 4]
```diff
@@ -977,6 +978,82 @@ func TestStakingSc_ExecuteIsStaked(t *testing.T) {
...
+		require.Nil(t, err)
+	}
+
+	for i := 10; i < 20; i++ {
+		idxStr := strconv.Itoa(i)
+		addr := []byte("addr" + idxStr)
+		doStake(t, stakingSmartContract, stakingAccessAddress, addr, addr)
+		checkIsStaked(t, stakingSmartContract, addr, addr, vmcommon.Ok)
+	}
+	requireRegisteredNodes(t, stakingSmartContract, eei, 14, 0)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking_test.go:1026]: This is the expected behavior. Once staking v4 is enabled, we can stake any number of nodes, without limit.

On `L1007` ou can see that after trying to stake 10 nodes, we expect to have: 4 staked nodes + 6 waiting.

Here, after `flagStakingV4` is enabled, waiting queue is removed => we all 14 nodes are staked

[the remark this answers is not in the document]
Comment (ssd04) [reply vm/systemSmartContracts/staking_test.go:1026]: ok :+1:  thanks

**Review by ssd04 — DISMISSED** [review]

File: vm/systemSmartContracts/staking.go [hunk 5]
```diff
@@ -659,12 +675,14 @@ func (s *stakingSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
...
 	}
 
-	addOneFromQueue := !s.flagCorrectLastUnjailed.IsSet() || s.canStakeIfOneRemoved()
-	if addOneFromQueue {
-		_, err = s.moveFirstFromWaitingToStaked()
-		if err != nil {
-			s.eei.AddReturnMessage(err.Error())
-			return vmcommon.UserError
+	if !s.flagStakingV4.IsSet() {
+		addOneFromQueue := !s.flagCorrectLastUnjailed.IsSet() || s.canStakeIfOneRemoved()
```
Comment (sasurobert) [inline vm/systemSmartContracts/staking.go:679]: this file will need one big cleanup - like stakingV2 separate file and stakingV4 separate. Add a task in JIRA and we will see later.

**Review by sasurobert — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**